scripts/components/RecentChanges/RecentChanges.js

```diff
@@ -1,175 +1,116 @@
 import React, { Component } from 'react'
-import SkeletalLoading from './SkeletalLoading'
-import RenderChangeSet from './RenderChangeSet'
+import cloneDeep from 'lodash.clonedeep'
+
+import { useScrollNearEnd } from '../../hooks'
+import SkeletonChangeSet from './SkeletonChangeSet'
+import ChangeSet from './ChangeSet'
 
 const HISTORY_ENDPOINT =
   window.location.host.endsWith('now.sh') ||
   process.env.NODE_ENV === 'development'
     ? 'https://api.mousey.app/v3/emoji/test/changes'
     : 'https://api.mousey.app/v3/emoji/blobs+community-blobs/changes'
 
-function historyTransform(historyData) {
+const HISTORY_EVENTS = [
+  'EMOJI_CREATE',
+  'EMOJI_RENAME',
+  'EMOJI_UPDATE',
+  'EMOJI_REMOVE',
+]
+
+function groupHistory(changeSets) {
   let changes = {}
-  for (const change of historyData) {
-    if (
-      ['EMOJI_CREATE', 'EMOJI_RENAME', 'EMOJI_UPDATE', 'EMOJI_REMOVE'].includes(
-        change.event
-      )
-    ) {
-      const date = new Date(change.changed_at)
-      const key = [
-        change.guild.id,
-        date.getFullYear(),
-        date.getMonth(),
-        date.getDate(),
-        date.getHours(),
-      ]
-      changes[key] = changes[key] ?? []
-      changes[key].push(change)
+
+  for (const changeSet of changeSets) {
+    if (!HISTORY_EVENTS.includes(changeSet.event)) {
+      continue
     }
-  }
 
-  return changes
-}
+    const date = new Date(changeSet.changed_at)
 
-export default class RecentChangesWrapper extends Component {
-  constructor(props) {
-    super(props)
+    const key = [
+      changeSet.guild.id,
+      date.getFullYear(),
+      date.getMonth(),
+      date.getDate(),
+      date.getHours(),
+    ].join('_')
 
-    this.state = {
-      changes: {},
-      fetching: false,
+    if (!changes[key]) {
+      changes[key] = []
     }
+
+    changes[key].push(changeSet)
   }
 
-  shouldComponentUpdate(nextProps, nextState, nextContext) {
-    const a = { ...this.state }
-    delete a.fetching
-    delete nextState.fetching
+  return changes
+}
 
-    return JSON.stringify(a) !== JSON.stringify(nextState)
-  }
+export default function RecentChanges() {
+  const [changeSets, setChangeSets] = React.useState(null)
```

Comment:
This has to do with TypeScript being unable to infer the types. The type is `object | null` but it can't figure that out. I check for `null` later on, so it shouldn't be much of a problem.